### PR TITLE
sentinel client application metrics persistent

### PR DESCRIPTION
**Pull request by YoungHu**
### Describe what this PR does / why we need it
add new feature with issue #88 

### Does this pull request fix one issue?
#88 

### Describe how you did it
save metrics data into mysql DB

### Describe how to verify it
check the mysql DB & sentinel-dashboard monitor page

### Special notes for reviews
1. no

### CODE REVIEW

File: sentinel-dashboard/pom.xml [hunk 1]
```diff
@@ -110,7 +115,21 @@
...
             <version>1.2.47</version>
         </dependency>
-
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>5.1.44</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
```
Comment (kimmking) [inline sentinel-dashboard/pom.xml]: nice, refer lombok can give more benefit.

File: sentinel-dashboard/src/main/java/com/taobao/csp/sentinel/dashboard/datasource/entity/MetricEntity.java [hunk 2]
```diff
@@ -92,132 +104,35 @@ public synchronized void addRtAndSuccessQps(double avgRt, Long successQps) {
...
-            ", resource='" + resource + '\'' +
-            ", passedQps=" + passedQps +
-            ", blockedQps=" + blockedQps +
-            ", successQps=" + successQps +
-            ", exception=" + exception +
-            ", rt=" + rt +
-            ", count=" + count +
-            ", resourceCode=" + resourceCode +
-            '}';
+                "id=" + id +
```
Comment (kimmking) [inline sentinel-dashboard/src/main/java/com/taobao/csp/sentinel/dashboard/datasource/entity/MetricEntity.java]: 也可以用lombok

File: sentinel-dashboard/src/main/resources/spring/spring-db.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans
+	xmlns="http://www.springframework.org/schema/beans"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd">
+
+	<!-- mysql data source -->
+	<bean id="mysqlDataSource" class="com.alibaba.druid.pool.DruidDataSource" init-method="init" destroy-method="close" primary="true">
+		<property name="driverClassName" value="com.mysql.jdbc.Driver" />
+		<property name="url" value="jdbc:mysql://xxx.xxx.xxx.xxx:3306/sentinel" />
```
Comment (kimmking) [inline sentinel-dashboard/src/main/resources/spring/spring-db.xml]: 改成localhost作为默认值比较好

File: sentinel-dashboard/src/main/resources/spring/spring-db.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		<property name="username" value="xxx" />
```
Comment (kimmking) [inline sentinel-dashboard/src/main/resources/spring/spring-db.xml]: 用root 123456

Comment (kimmking) [reply sentinel-dashboard/src/main/resources/spring/spring-db.xml]: 最好放到application.properties

File: sentinel-dashboard/src/main/java/com/taobao/csp/sentinel/dashboard/datasource/entity/MetricEntity.java (hunk 2 above)
Comment (YoungHu) [reply sentinel-dashboard/src/main/java/com/taobao/csp/sentinel/dashboard/datasource/entity/MetricEntity.java]: 嗯，看原来是写了，就想没必要改了

File: sentinel-dashboard/src/main/resources/spring/spring-db.xml (hunk 4 above)
Comment (YoungHu) [reply sentinel-dashboard/src/main/resources/spring/spring-db.xml]: 配置文件是个好建议，上来就搞了，没太在意这个细节

File: sentinel-dashboard/src/main/resources/spring/spring-db.xml (hunk 3 above)
Comment (YoungHu) [reply sentinel-dashboard/src/main/resources/spring/spring-db.xml]: 关于这个DB的默认值，我个人感觉不管写成啥都不合适。基本写的默认值每个人在使用的时候还是要去改一下。我个人喜欢把需要自己改的东西用`xxx`代替，哈哈。

File: sentinel-dashboard/src/main/resources/spring/spring-db.xml (hunk 4 above)
Comment (kimmking) [reply sentinel-dashboard/src/main/resources/spring/spring-db.xml]: default value means out-of-box use.

**Comment by sczyh30** [comment]
Hi, thanks for your contribution! Some considerations can be found in my comments of #50. 

Besides, in my opinion, using MySQL may not be a good idea for production when data grows fast. Maybe using time-series database (e.g. InfluxDB) is better.

**Comment by jialianglinjl** [comment]
Or clear the historical data periodically 





来自钉钉专属商务邮箱------------------------------------------------------------------
发件人：Eric Zhao<notifications@github.com>
日　期：2018年08月25日 18:58:12
收件人：alibaba/Sentinel<Sentinel@noreply.github.com>
抄　送：Subscribed<subscribed@noreply.github.com>
主　题：Re: [alibaba/Sentinel] sentinel client application metrics persistent (#89)

Hi, thanks for your contribution! Some considerations can be found in my comments of #50.
Besides, in my opinion, using MySQL may not be a good idea for production when data grows fast. Maybe using time-series database (e.g. InfluxDB) is better.

**Comment by sczyh30** [comment]
The persistence type and purge strategy depends on the data size, so in common scenarios RDBMS is okay. But I think the metrics persistence should be extensible and should not be only restricted to MySQL.

**Comment by YoungHu** [comment]
actually it's simple to change the the db from mysql to other db(like cratedb, TiDB) just need to change  the jdbc driver. on the other way, this is just mysql solution, anyone can reimplement it with other datasource.

**Comment by CarpenterLee** [comment]
This feature is necessary and your implementation seems fine. But there are several problems we should concern about: 
- The performance of Mysql, I recommend you to test the writing performance under more than 1000 resources per second;
- The expansibility of Mysql, assuming that 1000 resources are inserted into DB per second, meaning 86M lines per day. When data growing up, how to ensure the inserting performance and how we deal with the historical data?
- The pluggability. We want to ensoure the dashboard could work with little dependencies. So the metric persistence feature should be optional. 

I think this implementation will be fine in some specified conditions, but there are several problems to be considered.

**Comment by sczyh30** [comment]
@YoungHu Interfaces for metrics persistence are not settled down, so this PR might be pending until it has been completed. Suggestions are welcomed. Besides, we are interested in your practice about integrations with Grafana. Would you like to write a blog post about it (in Chinese)? You can also introduce your practice about metrics persistence for Sentinel Dashboard there.

**Comment by sczyh30** [comment]
Hi, our awesome repo is ready. You can put your implementation in your own repository, then link it in [Awesome Sentinel](https://github.com/alibaba/sentinel-awesome) with a brief description. It would be better if you could write a blog post about it.

**Comment by YoungHu** [comment]
ok, may done it in this week

**Comment by YoungHu** [comment]
this is the blog link: https://blog.csdn.net/huyong1990/article/details/82392386

**Comment by sczyh30** [comment]
As this has been added to Awesome Sentinel, closing the PR.

**Pull request closed without merging**